### PR TITLE
evaluator: support json_merge/json_type/json_unquote/json_extract

### PR DESCRIPTION
**Pull request by AndreMouche**
Hi,
     This PR support functions `json_merge`, `json_type`, `json_unquote`, `json_extract`, `json_insert`, `json_set`, `json_replace` in `evaluator`, the related PR in `tidb` is https://github.com/pingcap/tidb/pull/3492,
and the related source files in `tidb` are

- https://github.com/pingcap/tidb/blob/master/expression/builtin_json.go
- https://github.com/pingcap/tidb/blob/master/distsql/xeval/eval_json.go
- https://github.com/pingcap/tidb/blob/master/distsql/xeval/eval.go

@hhkbp2 @hicqu @lishihai9017 @andelf PTAL

### CODE REVIEW

File: src/util/codec/datum.rs [hunk 1]
```diff
@@ -273,7 +273,7 @@ impl Datum {
                 Json::String(String::from(data)).cmp(json)
             }
             _ => {
-                let data = format!("{:?}", *self);
+                let data = self.as_string().unwrap_or_default();
```
Comment (hhkbp2) [inline src/util/codec/datum.rs]: Is it suitable to use default value here?

Comment (AndreMouche) [inline src/util/codec/datum.rs]: yes, just as `tidb`(https://github.com/pingcap/tidb/blob/master/util/types/datum.go#L657) did. @hhkbp2

File: src/util/codec/datum.rs [hunk 2]
```diff
@@ -299,23 +299,32 @@ impl Datum {
         Ok(b)
     }
 
-    /// into_string convert self into a string.
-    /// source function name is `ToString`.
-    pub fn into_string(self) -> Result<String> {
-        let s = match self {
+    pub fn as_string(&self) -> Result<String> {
```
Comment (hhkbp2) [inline src/util/codec/datum.rs]: IMO it's better to use name `into_string`. For name convention in Rust, `as_xxx` is usually for returning a reference, such as, `as_bytes()`, `as_str()`, `as_ref()`, while `into_xxx` is for type conversion, such as, `into_bytes()`, `into_inner()`.

File: src/util/codec/datum.rs [hunk 3]
```diff
@@ -399,6 +408,59 @@ impl Datum {
...
+            _ => {
+                let s = try!(self.into_string());
+                Ok(Json::String(s))
+            }
+        }
+    }
+
+    /// into_json would convert Datum::Bytes(bs) into Json::String(bs)
+    /// and convert Datum::Null into Json::None.
+    /// This func would be used in json_unquote && json_modify
```
Comment (hhkbp2) [inline src/util/codec/datum.rs]: s/&&/and/

File: src/util/xeval/evaluator.rs [hunk 4]
```diff
@@ -225,13 +232,40 @@ impl Evaluator {
...
         Ok((left, right))
     }
 
+    fn eval_more_children(&mut self,
+                          ctx: &EvalContext,
+                          expr: &Expr,
+                          num: usize)
+                          -> Result<Vec<Datum>> {
+        let children = expr.get_children();
+        if children.len() < 2 {
```
Comment (hhkbp2) [inline src/util/xeval/evaluator.rs]: s/2/num/

File: src/util/xeval/evaluator.rs [hunk 5]
```diff
@@ -386,6 +420,91 @@ impl Evaluator {
...
+        // +------------------------------+
+        // | {"a":   "b"}                 |
+        // +------------------------------+
+        let json = try!(child.into_json());
+        let unquote_data = try!(json.unquote());
+        Ok(Datum::Bytes(unquote_data.into_bytes()))
+    }
+
+    fn eval_json_extract(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let children = try!(self.eval_more_children(ctx, expr, 2));
```
Comment (hhkbp2) [inline src/util/xeval/evaluator.rs]: `JSON_EXTRACT()` in mysql supports multiple paths.

File: src/util/xeval/evaluator.rs [hunk 6]
```diff
@@ -999,4 +1135,96 @@ mod test {
             }
         }
     }
+
+    fn build_byte_datums_expr(data: Vec<&[u8]>, tp: ExprType) -> Expr {
```
Comment (andelf) [inline src/util/xeval/evaluator.rs]: How about &[&[u8]]

File: src/util/codec/datum.rs [hunk 7]
```diff
@@ -1210,7 +1272,7 @@ mod test {
              Datum::Json(Json::Double(i32::MIN as f64)), Ordering::Equal),
             (b"hi".as_ref().into(), Datum::Json(Json::from_str(r#""hi""#).unwrap()),
              Ordering::Equal),
-            (Datum::Max,Datum::Json(Json::from_str(r#""MAX""#).unwrap()),Ordering::Equal),
+            (Datum::Max,Datum::Json(Json::from_str(r#""MAX""#).unwrap()),Ordering::Less),
```
Comment (andelf) [inline src/util/codec/datum.rs]: Space after comma

File: src/util/xeval/evaluator.rs [hunk 8]
```diff
@@ -386,6 +420,91 @@ impl Evaluator {
...
+        }
+    }
+
+    fn eval_json_modify(&mut self,
+                        ctx: &EvalContext,
+                        expr: &Expr,
+                        mt: ModifyType)
+                        -> Result<Datum> {
+        let children = try!(self.eval_more_children(ctx, expr, 2));
+        if children.len() & 1 == 0 {
```
Comment (hhkbp2) [inline src/util/xeval/evaluator.rs]: wrap a `is_even()` util function for this line.

**Comment by hhkbp2** [comment]
Rest LGTM

File: src/util/xeval/evaluator.rs [hunk 9]
```diff
@@ -386,6 +420,91 @@ impl Evaluator {
...
+    fn eval_json_merge(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let children = try!(self.eval_more_children(ctx, expr, 2));
+        if children.iter().any(|item| *item == Datum::Null) {
+            return Ok(Datum::Null);
+        }
+        let mut children = children.into_iter();
+        let first = try!(children.next().unwrap().cast_as_json());
+        let suffixes: Vec<Json> = try!(children.map(|item| item.cast_as_json())
+            .collect());
+        Ok(Datum::Json(first.merge(suffixes)))
```
Comment (andelf) [inline src/util/xeval/evaluator.rs]: how about Vec::split_off()?

File: src/util/xeval/evaluator.rs [hunk 10]
```diff
@@ -386,6 +420,91 @@ impl Evaluator {
...
+        let children = try!(self.eval_more_children(ctx, expr, 2));
+        if children.iter().any(|item| *item == Datum::Null) {
+            return Ok(Datum::Null);
+        }
+        let mut children = children.into_iter();
+        let json = try!(children.next().unwrap().cast_as_json());
+        let path_extrs: Vec<PathExpression> = try!(children.map(|item| item.into_json_path_expr())
+            .collect());
+        if let Some(data) = json.extract(&path_extrs) {
+            Ok(Datum::Json(data))
```
Comment (andelf) [inline src/util/xeval/evaluator.rs]: same pattern as above. :)

File: src/util/codec/datum.rs [hunk 11]
```diff
@@ -399,6 +408,59 @@ impl Datum {
...
+            Datum::Null => Ok(Json::None),
+            Datum::Bytes(bs) => {
+                let s = try!(String::from_utf8(bs));
+                Ok(Json::String(s))
+            }
+            _ => self.cast_as_json(),
+        }
+    }
+
+    pub fn into_json_path_expr(self) -> Result<PathExpression> {
```
Comment (andelf) [inline src/util/codec/datum.rs]: since following code doesn't benefit from move self, `&self` is better.

File: src/util/codec/datum.rs [hunk 12]
```diff
@@ -399,6 +408,59 @@ impl Datum {
...
+        let v = match self {
+            Datum::Bytes(bs) => try!(String::from_utf8(bs)),
+            _ => String::default(),
```
Comment (andelf) [inline src/util/codec/datum.rs]: safe to treat any other Datum as empty expression? (

File: src/util/xeval/evaluator.rs [hunk 13]
```diff
@@ -225,13 +232,40 @@ impl Evaluator {
...
         let (left_expr, right_expr) = try!(self.get_two_children(expr));
         let left = try!(self.eval(ctx, left_expr));
         let right = try!(self.eval(ctx, right_expr));
         Ok((left, right))
     }
 
+    fn eval_more_children(&mut self,
+                          ctx: &EvalContext,
+                          expr: &Expr,
+                          num: usize)
```
Comment (lishihai9017) [inline src/util/xeval/evaluator.rs]: the name doesn't express at least

File: src/util/xeval/evaluator.rs (hunk 9 above)
Comment (lishihai9017) [inline src/util/xeval/evaluator.rs]: but it brings copy action although it make code clear.

**Comment by lishihai9017** [comment]
Rest LGTM

File: src/util/xeval/evaluator.rs (hunk 13 above)
Comment (AndreMouche) [inline src/util/xeval/evaluator.rs]: I'd like to express more than `num` children, any better suggestion? @lishihai9017

**Comment by siddontang** [comment]
IMO, you should split your PR into at lease 4 PRs, not one.

File: src/util/xeval/evaluator.rs (hunk 9 above)
Comment (andelf) [inline src/util/xeval/evaluator.rs]: both OK for me. 😆

File: src/util/xeval/evaluator.rs [hunk 14]
```diff
@@ -460,6 +579,10 @@ fn check_in(ctx: &EvalContext, target: Datum, value_list: &[Datum]) -> Result<bo
     Ok(pos.is_ok())
 }
 
+fn is_even(n: i64) -> bool {
```
Comment (andelf) [inline src/util/xeval/evaluator.rs]: #[inline]

**Comment by AndreMouche** [comment]
This PR was translated from `tidb`'s PR (https://github.com/pingcap/tidb/pull/3492). I prefer to put it in one PR like `tidb` did, so it would be more clear for the reviewers. @siddontang

**Comment by siddontang** [comment]
@AndreMouche 

One PR, one thing. Don't break the rule even the associated PR does it. 

@hicqu , please follow the rule later.

File: src/util/xeval/evaluator.rs [hunk 15]
```diff
@@ -386,6 +420,91 @@ impl Evaluator {
...
+                        ctx: &EvalContext,
+                        expr: &Expr,
+                        mt: ModifyType)
+                        -> Result<Datum> {
+        let children = try!(self.eval_more_children(ctx, expr, 2));
+        if is_even(children.len() as i64) {
+            return Err(Error::Expr(format!("expect odd number operands, got {}", children.len())));
+        }
+
+        if children.iter().any(|item| *item == Datum::Null) {
```
Comment (hicqu) [inline src/util/xeval/evaluator.rs]: Here the logic is wrong. See: https://github.com/pingcap/tidb/pull/3562.
Sorry for that, I made a mistake.

**Pull request closed without merging**